syntax = "proto3";

//Topics
service TopicService {
    rpc create(TopicRequest) returns (TopicResponse) {}
    rpc delete(TopicRequest) returns (TopicResponse) {}
    rpc listTopics(Empty) returns (TopicResponseList) {}
    rpc subscribeTopic(TopicRequest) returns (TopicResponse) {}
    rpc publishTopic(TopicRequestMessage) returns (TopicResponse) {}
    rpc consumeTopic(TopicRequest) returns (TopicResponseMessage) {}
}

message TopicRequest {
    string key = 1;
    string topic_name = 2;
}

message TopicResponse {
    sint32 code = 1;  
}

message TopicResponseList {
    sint32 code = 1;
    repeated string topics = 2;
}

message TopicRequestMessage {
    string key = 1;
    string topic_name = 2;
    string message = 3;
}

message TopicResponseMessage {
    sint32 code = 1;
    string message = 2;
}

//Queues
service QueueService {
    rpc create(QueueRequest) returns (QueueResponse) {}
    rpc delete(QueueRequest) returns (QueueResponse) {}
    rpc listQueues(Empty) returns (QueueResponseList) {}
    rpc subscribeQueue(QueueRequest) returns (QueueResponse) {}
    rpc publishQueue(QueueRequestMessage) returns (QueueResponse) {}
    rpc consumeQueue(QueueRequest) returns (QueueResponseMessage) {}
}

message QueueRequest {
    string key = 1;
    string queue_name = 2;
}

message QueueResponse {
    sint32 code = 1;  
}

message QueueResponseList {
    sint32 code = 1;
    repeated string queues = 2;
}

message QueueRequestMessage {
    string key = 1;
    string queue_name = 2;
    string message = 3;
}

message QueueResponseMessage {
    sint32 code = 1;
    string message = 2;
}

message Empty {}